src/boostvqe/training_utils.py

```diff
@@ -319,3 +321,38 @@ def parameter_shift(
 
     circuit.set_parameters(original)
     return float(generator_eigenval * (forward - backward))
+
+
+def quantum_derivative_via_shifts(
+    hamiltonian, circuit, parameter_index, exec_backend, nshots=None
+):
+    target_gate = circuit.associate_gates_with_parameters()[parameter_index]
+    target_qubits = target_gate.qubits
+    parameter_value = target_gate.parameters[0]
+
+    if target_gate.name in ["rx", "ry", "rz"]:
+        return parameter_shift(
+            hamiltonian, circuit, parameter_index, exec_backend, nshots
+        )
+    elif target_gate.name == "rbs":
+        unrolled_rbs_circuit = Circuit(circuit.nqubits)
+        for g in circuit.queue:
+            if g != target_gate:
+                unrolled_rbs_circuit.add(g)
+            else:
+                unrolled_rbs_circuit.add(gates.H(target_qubits[0]))
+                unrolled_rbs_circuit.add(gates.CNOT(target_qubits[0], target_qubits[1]))
+                unrolled_rbs_circuit.add(gates.H(target_qubits[1]))
+                unrolled_rbs_circuit.add(gates.RY(target_qubits[0], parameter_value))
+                unrolled_rbs_circuit.add(gates.RY(target_qubits[1], -parameter_value))
+                unrolled_rbs_circuit.add(gates.H(target_qubits[1]))
```

Comment:
I only saw this now, but you could remove these two :math:`H` gates and flip the sign of the second :math:`RY` from negative to positive and it is also a valid decomposition.

Reply:
> I only saw this now, but you could remove these two H gates and flip the sign of the second R Y from negative to positive and it is also a valid decomposition.

I'll update the default decomposition in `qibo` to the one with less qubits.

Reply:
Yes, it's true! Don't think it is affecting that much the computation, though.